File: src/main/java/de/mikeyllp/miniGamesV4/listeners/PlayerJoinQuitListener.java
package de.mikeyllp.miniGamesV4.listeners;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.List;
import java.util.Map;

import static de.mikeyllp.miniGamesV4.game.hideandseek.storage.HideAndSeekGameGroups.*;
import static de.mikeyllp.miniGamesV4.game.hideandseek.utils.removePlayersHideAndSeek.playerRemove;
import static de.mikeyllp.miniGamesV4.game.rps.RPSGame.inGameStatus;
import static de.mikeyllp.miniGamesV4.game.rps.RPSGame.removePlayersFromList;
import static de.mikeyllp.miniGamesV4.utils.MessageUtils.sendCustomMessage;

public class PlayerJoinQuitListener implements Listener {

    private final JavaPlugin plugin;

    public PlayerJoinQuitListener(JavaPlugin plugin) {
        this.plugin = plugin;
    }

    //removes the player from the game when he leaves the server
    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        Player player = event.getPlayer();

        // Check if the player was in RPS game
        if (inGameStatus.containsKey(player)) {
            sendCustomMessage(inGameStatus.get(player), "<red>" + player.getName() + " hat das Spiel verlassen!</red>");
            removePlayersFromList(player, inGameStatus.get(player));
        }

        // Removes The Player from the listUntilX
        listUntilX.removeIf(value -> value.equals(player));
        playerRemove(player, "disconnected", plugin);

    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        Player player = event.getPlayer();

        if (!gameState.isEmpty()) {
            for (Map.Entry<String, List<Player>> entry : gameGroup.entrySet()) {
                List<Player> groupPlayers = entry.getValue();
                for (Player groupPlayer : groupPlayers) {
                    groupPlayer.hidePlayer(plugin, player);
                }
                for (Player outsider : Bukkit.getOnlinePlayers()) {
                    if (groupPlayers.contains(outsider)) continue;
                    for (Player hidden : groupPlayers) {
                        outsider.hidePlayer(plugin, hidden);
                    }
                }
            }
        }
    }
}

File: src/main/java/de/mikeyllp/miniGamesV4/listeners/PlayerJoinQuitListener.kt
package de.mikeyllp.miniGamesV4.listeners

import de.mikeyllp.miniGamesV4.games.hideandseek.storage.HideAndSeekGameGroups
import de.mikeyllp.miniGamesV4.games.hideandseek.utils.removePlayersHideAndSeek
import de.mikeyllp.miniGamesV4.games.rps.RPSGame
import de.mikeyllp.miniGamesV4.utils.MessageUtils
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.plugin.java.JavaPlugin

class PlayerJoinQuitListener(private val plugin: JavaPlugin) : Listener {
    //removes the player from the game when he leaves the server
    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        val player = event.getPlayer()

        // Check if the player was in RPS game
        if (RPSGame.Companion.inGameStatus.containsKey(player)) {
            MessageUtils.sendMessage(
                RPSGame.Companion.inGameStatus.get(player),
                "<red>" + player.getName() + " hat das Spiel verlassen!</red>"
            )
            RPSGame.Companion.removePlayersFromList(player, RPSGame.Companion.inGameStatus.get(player))
        }

        // Removes The Player from the listUntilX
        HideAndSeekGameGroups.Companion.listUntilX.removeIf { value: Player? -> value == player }
        removePlayersHideAndSeek.playerRemove(player, "disconnected", plugin)
    }

    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        val player = event.getPlayer()
        println(player.locale())

        if (!HideAndSeekGameGroups.Companion.gameState.isEmpty()) {
            for (entry in HideAndSeekGameGroups.Companion.gameGroup.entries) {
                val groupPlayers: MutableList<Player> = entry.value
                for (groupPlayer in groupPlayers) {
                    groupPlayer.hidePlayer(plugin, player)
                }
                for (outsider in Bukkit.getOnlinePlayers()) {
                    if (groupPlayers.contains(outsider)) continue
                    for (hidden in groupPlayers) {
                        outsider.hidePlayer(plugin, hidden)
                    }
                }
            }
        }
    }
}
